pep-0702.rst

```diff
@@ -0,0 +1,278 @@
+  methods are considered.
+* A class may use the ``@override`` decorator from :pep:`698` to assert that
+  its method overrides a base class method, but the base class method is
+  deprecated.
+
+As these scenarios appear complex and relatively unlikely to come up in practice,
+this PEP does not mandate that type checkers detect them.
+
+Example
+-------
+
+As an example, consider this library stub named ``library.pyi``:
+
+.. code-block:: python
+
+   from typing import deprecated
+
+   @deprecated("Use Spam instead")
+   class Ham: ...
+
+   @deprecated("It is pining for the fiords")
+   def norwegian_gray(x: int) -> int: ...
+
+   @deprecated("Only str will be allowed")
+   @overload
+   def foo(x: int) -> str: ...
+   @overload
+   def foo(x: str) -> str: ...
+
+Here is how type checkers should handle usage of this library:
+
+.. code-block:: python
+
+   from library import Ham  # error: Use of deprecated class Ham. Use Spam instead.
+
+   import library
+
+   library.norwegian_gray(1)  # error: Use of deprecated function norwegian_gray. It is pining for the fiords.
+   map(library.norwegian_gray, [1, 2, 3])  # error: Use of deprecated function norwegian_gray. It is pining for the fiords.
+
+   library.foo(1)  # error: Use of deprecated overload for foo. Only str will be allowed.
+   library.foo("x")  # no error
+
+Runtime behavior
+----------------
+
+At runtime, the decorator sets an attribute ``.__deprecated__`` on the decorated
+object. The value of the attribute is the message passed to the decorator.
+The decorator returns the original object. Notably, it does not issue a runtime
+``DeprecationWarning``.
+
+Type checker behavior
+---------------------
+
+This PEP does not specify exactly how type checkers should present deprecation
+diagnostics to their users. However, some users (e.g., application developers
+targeting only a specific version of Python) may not care about deprecations,
+while others (e.g., library developers who want their library to remain
+compatible with future versions of Python) would want to catch any use of
+deprecated functionality in their CI pipeline. Therefore, it is recommended
+that type checkers provide configuration options that cover both use cases.
+
+Backwards compatibility
+=======================
+
+Creating a new decorator poses no backwards compatibility concerns.
+As with all new typing functionality, the ``@deprecated`` decorator
+will be added to the ``typing_extensions`` module, enabling its use
+in older versions of Python.
+
+How to teach this
+=================
+
+For users who encounter deprecation warnings in their IDE or type
+checker output, the messages they receive should be clear and self-explanatory.
+Usage of the ``@deprecated`` decorator will be an advanced feature
+mostly relevant to library authors. The decorator should be mentioned
+in relevant documentation (e.g., :pep:`387` and the :exc:`DeprecationWarning`
```

Comment:
PEP 387 (PEP-387) specifies a normative process for Python's own development, rather than being itself (user) documentation. Maybe it would be worth being explicit here what change this PEP is recommending to that one (e.g. to require `@deprecated` when possible as part of the Python deprecation process, alongside `DeprecationWarning`/`warnings._deprecated` and `.. deprecated::`/`.. deprecated-removed::` in the docs?)

Reply:
I think we should, but that hardly fits under "How to teach this" any more; I'll add a new section to the spec instead.